oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java

```diff
@@ -772,6 +784,128 @@ public UserManager getUserManager() throws RepositoryException {
+                        }
+                    };
+                }
+                return null;
+            }
+        });
+    }
+
+    @Nonnull
+    @Override
+    public Binary completeHttpUpload(String uploadToken) throws RepositoryException {
+        return sd.perform(new ReadOperation<Binary>("completeHttpUpload") {
+
+            @Nonnull
+            @Override
+            public Binary perform() throws RepositoryException {
+
+                HttpBlobProvider httpBlobProvider = getHttpBlobProvider();
+                if (httpBlobProvider == null) {
+                    throw new RepositoryException("HTTP binary upload not supported");
+                }
+                Blob blob = httpBlobProvider.completeHttpUpload(uploadToken);
+                if (blob == null) {
+                    throw new RepositoryException("HTTP binary upload could not be completed");
+                }
+
+                ValueFactoryImpl vf = (ValueFactoryImpl) sessionContext.getValueFactory();
+                return vf.createBinary(blob);
+            }
+        });
+    }
+
+    @Nullable
+    @Override
+    public URL getHttpDownloadURL(Binary binary) {
+        return sd.safePerformNullable(new ReadOperation<URL>("getHttpDownloadURL") {
+
+            @Nullable
+            @Override
+            public URL performNullable() throws RepositoryException {
+                HttpBlobProvider httpBlobProvider = getHttpBlobProvider();
+                if (httpBlobProvider == null) {
+                    return null;
+                }
+                if (binary instanceof ReferenceBinary) {
+                    String blobReference = ((ReferenceBinary) binary).getReference();
+                    if (null == blobReference) {
+                        // Binary is inlined, we cannot return a URL for it
+                        return null;
+                    }
+                    Blob blob = sd.getRoot().getBlob(blobReference);
+                    if (blob != null) {
+                        // TODO: how to check if this blob is actually a blob from that blob store?
+                        //       and not an inlined SegmentBlob or from another data store?
+                        //       a) DS could check for existence before presigning, but that'll be costly (network request vs. none)
+                        //       b) looking at the blob id will not help as that is not specific per DS
+                        //       c) alternatively we go the route of URLReadableBlob from patch v1 and add this to
+                        //          the blob impls itself, essentially changing the approach from provider to this, but only
+                        //          for HttpBlobProvider and below - for the JCR level API HttpBinaryProvider we should stick
+                        //          to that single simple interface for write + read IMO
+                        //       d) maybe the secure reference already helps?
```

Comment:
(Just so that this does not get forgotten) We need to solve this "how to check if this blob is actually a blob from that blob store?"

A possible test might be using a `CompositeDataStore` with two S3 data stores under the hood. The problem would surface there that in the `CompositeDataStore` you'd have to implement the `HttpDataRecordProvider` but based on the reference would not know which underlying datastore the blob belongs to (unless I am mistaken).

I think there might be other situations where it could pop up.